lucet-runtime/lucet-runtime-internals/src/future.rs

```diff
@@ -150,91 +169,179 @@ impl InstanceHandle {
-                // `resume_val`, send it back to the guest ctx and continue
-                // running:
-                self.resume_with_val_impl(
-                    resume_val
-                        .take()
-                        .expect("is_yielded implies resume_value is some"),
-                    true,
-                    runtime_bound,
-                )
-            } else if self.is_bound_expired() {
-                self.resume_bounded(
-                    runtime_bound.expect("should have bound if guest had expired bound"),
-                )
-            } else {
+    /// Returns a `RunAsync` that will asynchronously execute the guest instnace.
+    fn run_async_internal<'a>(&'a mut self, func: FunctionHandle, args: &'a [Val]) -> RunAsync<'a> {
+        RunAsync {
+            inst: self,
+            inst_count_bound: DEFAULT_INST_COUNT_BOUND,
+            state: RunAsyncState::Start(func, args),
+        }
+    }
+
+    /// Returns a `RunAsync` that will immediately fail with the given error, without executing the guest instance.
+    fn run_async_failed<'a>(&'a mut self, err: Error) -> RunAsync<'a> {
+        RunAsync {
+            inst: self,
+            inst_count_bound: DEFAULT_INST_COUNT_BOUND,
+            state: RunAsyncState::Failed(err),
+        }
+    }
+}
+
+/// A future implementation that enables running a guest instance which can call async hostcalls.
+///
+/// Lucet hostcalls are synchronous `extern "C" fn` functions called from WebAssembly. In that
+/// context, we cannot use `.await` directly because the hostcall is not `async`. While we could
+/// block on an executor such as `futures::executor::block_on`, that would block the OS thread,
+/// preventing us from running other async tasks while awaiting. This means an application will need more
+/// threads than otherwise would be necessary.
+///
+/// `RunAsync` allows the guest to call async hostcalls just as if the guest had called the async function
+/// and immediately `.await`ed it.
+///
+/// To define a async hostcall, simply add the async modifier to a hostcall definition:
+///
+/// ```ignore
+/// #[lucet_hostcall]
+/// #[no_mangle]
+/// pub async fn hostcall_async(vmctx: &Vmctx) {
+///    foobar().await
+/// }
+/// ```
+///
+/// Note: Async hostcalls may only be used if `Instance::run_async` was used to run the VM,
+/// otherwise it will terminate the instance with `TerminationDetails::BlockOnNeedsAsync`.
+///
+/// Behind the scenes, lucet polls the future from within the guest execution context. If the future is not immediately ready,
+/// the instance will yield and return control to the async executor. Later, when the future is ready to make progress,
+/// the async executor will return to the guest context, where lucet will poll the future to completion.
+///
+/// Just like `.await`, there is no overhead for futures that are immediately ready (such as `async { 5 }`).
```

Comment:
Slight clarity improvement: "There is no additional return to the executor loop when an async hostcall returns a future that is immediately ready. This behavior is the same as implemented by the `.await` operator."